collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -552,235 +730,177 @@ func (r *Receiver) flushSender(serverStream anyStreamServer, pendingCh <-chan ba
 		if r.Logs() == nil {
-			return status.Error(codes.Unimplemented, "logs service not available")
+			return status.Error(codes.Unimplemented, "logs service not available"), nil, 0, 0
 		}
 		var sizer plog.ProtoMarshaler
-		var numLogs int
-		ctx = r.obsrecv.StartLogsOp(ctx)
 
 		data, err := arrowConsumer.LogsFrom(records)
-		if err != nil {
-			err = consumererror.NewPermanent(err)
-		} else {
+		if err == nil {
 			for _, logs := range data {
-				items := logs.LogRecordCount()
-				sz := int64(sizer.LogsSize(logs))
-
-				r.recvInFlightBytes.Add(ctx, sz)
-				r.recvInFlightItems.Add(ctx, int64(items))
-				numLogs += items
-				uncompSize += sz
-				err = multierr.Append(err,
-					r.Logs().ConsumeLogs(ctx, logs),
-				)
-			}
-
-			acquireErr := r.acquireAdditionalBytes(ctx, uncompSize, response, sizeHeaderFound)
-
-			if acquireErr != nil {
-				err = multierr.Append(err, acquireErr)
-				// if acquireAdditionalBytes() failed then the previously acquired bytes were already released i.e. nothing to releaes.
-				return err
+				numItems += logs.LogRecordCount()
+				uncompSize += int64(sizer.LogsSize(logs))
 			}
-
-			// entire request has been processed, decrement counter.
-			r.recvInFlightBytes.Add(ctx, -uncompSize)
-			r.recvInFlightItems.Add(ctx, int64(-numLogs))
 		}
-		r.obsrecv.EndLogsOp(ctx, streamFormat, numLogs, err)
-		return err
+		retData = data
+		retErr = err
 
 	case arrowpb.ArrowPayloadType_SPANS:
 		if r.Traces() == nil {
-			return status.Error(codes.Unimplemented, "traces service not available")
+			return status.Error(codes.Unimplemented, "traces service not available"), nil, 0, 0
 		}
 		var sizer ptrace.ProtoMarshaler
-		var numSpans int
-		ctx = r.obsrecv.StartTracesOp(ctx)
 
 		data, err := arrowConsumer.TracesFrom(records)
-		if err != nil {
-			err = consumererror.NewPermanent(err)
-		} else {
+		if err == nil {
 			for _, traces := range data {
-				items := traces.SpanCount()
-				sz := int64(sizer.TracesSize(traces))
+				numItems += traces.SpanCount()
+				uncompSize += int64(sizer.TracesSize(traces))
+			}
+		}
+		retData = data
+		retErr = err
+
+	default:
+		retErr = ErrUnrecognizedPayload
+	}
 
-				r.recvInFlightBytes.Add(ctx, sz)
-				r.recvInFlightItems.Add(ctx, int64(items))
+	return retErr, retData, numItems, uncompSize
+}
 
-				numSpans += items
-				uncompSize += sz
-				err = multierr.Append(err,
-					r.Traces().ConsumeTraces(ctx, traces),
-				)
-			}
+// consumeData invokes the next pipeline consumer for a received batch of data.
+// it uses the standard OTel collector instrumentation (receiverhelper.ObsReport).
+//
+// if any errors are permanent, returns a permanent error.
+func (r *Receiver) consumeData(ctx context.Context, data any, flight *inFlightData) error {
+	var anyPermanent bool
+	var wholeErr error
+	type unwrapper interface{ Unwrap() error }
+
+	oneOp := func(err error) {
+		if consumererror.IsPermanent(err) {
+			anyPermanent = true
+			err = err.(unwrapper).Unwrap()
```

Comment:
Why are we unwrapping the error?

Reply:
This was unnecessary -- I had to remind myself that errors.Unwrap() can take two forms, one is `Unwrap() []error` and that is how `IsPermanent()` works on multierr. I've removed the extra cruft.